File: serial_dds_gateway（实物通信）/can_id_codec.py
# 用途：RS02 USB-CAN 串口帧中 29 位 CAN ID 与 4 字节扩展原始字段的互转辅助函数。
# 说明：灵足旧工具暴露的是移位后的 raw 字段，右移 shift_right 位后取低 29 位即为真实 CAN ID。
"""CAN-ID 转换辅助函数（RS02 串口 USB-CAN 成帧）。"""

from __future__ import annotations


def ext_raw_to_can29(ext_raw_4b: bytes, shift_right: int = 3) -> int:
    """将串口 4 字节扩展字段转换为真实 29 位 CAN ID。"""
    if len(ext_raw_4b) != 4:
        raise ValueError("ext_raw_4b 必须为 4 字节")
    ext_u32 = int.from_bytes(ext_raw_4b, byteorder="big", signed=False)
    can29 = (ext_u32 >> shift_right) & 0x1FFFFFFF
    return can29


def can29_to_ext_raw(can29: int, shift_left: int = 3) -> bytes:
    """将真实 29 位 CAN ID 转换为串口 4 字节扩展字段。"""
    can29 &= 0x1FFFFFFF
    ext_u32 = (can29 << shift_left) & 0xFFFFFFFF
    return ext_u32.to_bytes(4, byteorder="big", signed=False)

File: serial_dds_gateway（实物通信）/src/can_id_codec.cpp
#include "can_id_codec.hpp"

namespace serial_dds_gateway {

// 4 字节扩展 raw 字段拼接为 32 位无符号整数，再右移 shift_right 位取低 29 位得到真实 CAN ID。
uint32_t ext_raw_to_can29(const std::array<uint8_t, 4>& ext_raw, int shift_right) {
  const uint32_t ext_u32 = (static_cast<uint32_t>(ext_raw[0]) << 24) | (static_cast<uint32_t>(ext_raw[1]) << 16) |
                           (static_cast<uint32_t>(ext_raw[2]) << 8) | static_cast<uint32_t>(ext_raw[3]);
  return (ext_u32 >> shift_right) & 0x1FFFFFFF;
}

// 29 位 CAN ID 左移 shift_left 位后拆回 4 字节大端 raw 字段。
std::array<uint8_t, 4> can29_to_ext_raw(uint32_t can29, int shift_left) {
  const uint32_t ext_u32 = ((can29 & 0x1FFFFFFF) << shift_left);
  return {
      static_cast<uint8_t>((ext_u32 >> 24) & 0xFF),
      static_cast<uint8_t>((ext_u32 >> 16) & 0xFF),
      static_cast<uint8_t>((ext_u32 >> 8) & 0xFF),
      static_cast<uint8_t>(ext_u32 & 0xFF),
  };
}

}  // 命名空间 serial_dds_gateway

